### PR TITLE
Gutenboarding: Calypso-Free Launch Button Injection

### LINKED ISSUE #46839: Focused Launch: set up A/B testing
Currently, the Focused Launch Flow modal is enabled via the temporary `gutenboarding/focused-launch-flow` flag.

The goal is to swap the flag with A/B testing logic

**Acceptance Criteria:**

- [x] After creating a site using `/start` flow, when clicking _Launch_ button in the editor, 50% of users should be redirected to `/start/launch-site` and 50% should start the Focused launch flow with no redirect.
- [x] ~When enabling `gutenboarding/focused-launch-flow` in production we should check where it's being used. For example, in [openCheckout](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/utils.ts#L29-L35) we should rather use `!isGutenboarding` to identify the launch flow we're in.~ Now uses ENABLE_FOCUSED_LAUNCH in backend flag.
- [x] Not going through Calypso iframe would be good in the context of https://github.com/Automattic/wp-calypso/issues/47803 so A/B test code could be run in PHP as described in the recommendation below

**Hint:**

We could follow the pattern already implemented by @Automattic/ganon by adding a filter and get the A/B test variations on the backend:
- Server-side A/B test reference: https://github.com/Automattic/wp-calypso/pull/47896
- Server-side launch injection: D53680-code
- A/B test code: https://github.com/Automattic/wp-calypso/pull/49004 D56270-code
- Experiment: pbxNRc-Ag-p2

### PR DESCRIPTION
**Pull request by yansern**
## Changes proposed in this Pull Request

**TLDR Version:** A majority of this PR's changes is related to the overhaul of the Launch button injection where most logic to determine which launch flow to load is now moved over to the server-side.

1. **Removed `editor-gutenboarding-launch`. Launch button moved to `editor-site-launch`.** 
Previously this was responsible for injecting launch button on to the top bar. This is now moved over to `editor-site-launch` under `editor-site-launch/src/launch-button/index.ts` to streamline things.

2. **Three (3) separate compilation entry points for `editor-site-launch`.**  
Previously both step-by-step and focused launch are bundled into a single build, now they are splitted out to ensure only the necessary script is loaded (always nice to shave off on bundle size). The `package.json` file was modified to build the following: 
    * `launch-button.ts` => `dist/launch-button.js`
    * `step-by-step-launch.ts` => `dist/step-by-step-launch.js`
    * `focused-launch.ts` => `dist/focused-launch.js`

3. **Server-side determines whether to load Step-By-Step, Focused or Redirect launch flow.**  
This can be seen at D55512-code. When calling `apply_filters('a8c_enable_editor_site_launch')` on the server side, it will return an object containing `site_slug`, `launch_url` and `launch_flow: "gutenboarding-launch" | "focused-launch" | "launch-site"`.

4. **Launch button no longer rely on `window.calypsoifyGutenberg`. It relies by `window.wpcomEditorSiteLaunch` object injected by server-side `editor-site-launch/index.php` when enqueuing scripts.**
This means launch button / launch flow can run without Calypso's iframe. The `window.wpcomEditorSiteLaunch` contains the same data as mentioned in point 3.

**Other considerations:**
1. ~~This would render the client-side `create/focused-launch-flow` feature flag useless but I haven't look into how to hide behind feature flag on server-side.~~ We currently still supporting `create/focused-launch-flow` feature flag but will soon transition to using Explat and variation assignment for testing.
2. There are other places (outside of our launch codes) still relying on data from `calypsoifyGutenberg`, especially the `calypsoifyGutenberg?.isGutenboarding` flag. So any `calypsoifyGutenberg` related cleanup to be pushed to another PR here https://github.com/Automattic/wp-calypso/pull/49067.

## Testing instructions

**Setting up**
1. Apply patch D55512-code.
2. Run `yarn dev --sync` on `apps\editing-toolkit`.

**Test Step-By-Step Launch**
1. Create a **Gutenboarding site** via `/new`.
2. Sandbox the newly created site.
3. Go to the block editor and click on the **Launch** button.
4. It should show **Step-By-Step Launch**.

**Test Focused Launch**
1. Create a **Control site** via `/start`.
2. Sandbox the newly created site.
3. Go to the `focused_launch_flow` experiment in ExPlat/Abacus and click `treatment` to assign this A/B variation to your WP user.
4. Go to the block editor and click on the **Launch** button.
5. It should show the **Focused Launch**.

**Test Redirect Launch**
1. You can use the same Control site above.
2. Go to the `focused_launch_flow` experiment in ExPlat/Abacus and click `control` to assign this A/B variation to your WP user.
3. Go to the block editor and click on the **Launch** button.
4. It should show **redirect to `/start/launch-site`**.

Fixes #46839 #47803

### CODE REVIEW

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+		launchButton.href = launchUrl;
+		launchButton.target = '_top';
+		launchButton.className = 'editor-gutenberg-launch__launch-button components-button is-primary';
+		launchButton.addEventListener( 'click', ( e: Event ) => {
+			// Prevent default behaviour
+			e.preventDefault();
+
+			recordTracksEvent( 'calypso_newsite_editor_launch_click', {
+				flow: launchFlow,
+			} );
```
Comment (yansern) [inline apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts]: @razvanpapadopol Take note that this previously has the params `is_new_flow` and `is_experimental`.
`is_new_flow` holds the value `true` when using step-by-step launch.
`is_experimental` holds the value `true` when feature picker is enabled, it is now enabled for all anyway.

I'm not sure what relevant data to pass in, but the new `flow` param can contain either`step-by-step`, `focused` or `redirect`.

Comment (razvanpapadopol) [reply apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts]: We should add the following props:
* `is_new_site`: we could return also `is_gutenboarding` from D55512-code
* `launch_flow`: which can be `gutenboarding-launch` for `step-by-step`, `focused-launch` or `launch-site` for `redirect`.
* `is_in_iframe`: eg of [utility function](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/block-inserter-modifications/contextual-tips/utils.js#L6-L12) for this

Check 431-gh-Automattic/tracks-events-registration for details.

**Review by alshakero — APPROVED** [review]
I liked this design much more! The testing steps I used differ a bit from the PR description.

1. Step by step. Works as described.
2. Focused Launch. I had to add ENABLE_FOCUSED_LAUNCH to sandbox.php
3. Redirect. Works as described.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 /**
  * External dependencies
  */
+import 'a8c-fse-common-data-stores';
```
Comment (yansern) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js:4]: This line is added because after a site is launched, the `editor-site-launch` module no longer loads.

The `wpcom-block-editor-nux` uses the site data store but it didn't import this `a8c-fse-common-data-stores` previously because it has always relied on `editor-site-launch` module to load it.

However, this `a8c-fse-common-data-stores`, also loads other unnecessary data store such is domains, plans, launch that is not relevant to `wpcom-block-editor-nux`. This is something to optimize for the future.

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import { STEP_BY_STEP_LAUNCH_FLOW, FOCUSED_LAUNCH_FLOW, REDIRECT_LAUNCH_FLOW } from '../constants';
+import './styles.scss';
+
+let handled = false;
+
+const launchFlowTracksSlug: Record< string, string > = {
+	[ STEP_BY_STEP_LAUNCH_FLOW ]: 'gutenboarding-launch',
+	[ FOCUSED_LAUNCH_FLOW ]: 'focused-launch',
+	[ REDIRECT_LAUNCH_FLOW ]: 'launch-site',
+};
```
Comment (yansern) [inline apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts]: Was wondering if you wanna streamline this or leave as is? Basically it looks like this atm:
```
{
"step-by-step": "gutenboarding-launch", 
"focused": "focused-launch",
"redirect": "launch-site"
}
```

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts (hunk 1 above)
Comment (yansern) [reply apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts]: Does `is_in_editor` hold the value`true` when on Calypso and `false` when on WP-Admin? (That's what is in the current code.)

Also, see https://github.com/Automattic/wp-calypso/pull/49004#discussion_r563685452.

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts (hunk 3 above)
Comment (razvanpapadopol) [reply apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts]: Ideally, we could initialize the `window.wpcomEditorSiteLaunch.launchFlow` property directly with the values in this object instead of keys so we don't need a map anymore.

Then we could change https://github.com/Automattic/wp-calypso/pull/49004/files#diff-5ca2166fe3b2f3927d78bb85557a2582921420cc4838e80c276670d324ee504dR62 to
```
launch_flow: launchFlow,
```

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts (hunk 1 above)
Comment (razvanpapadopol) [reply apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts]: @yansern those are the correct values. Sorry for misspelling the property name. It should be `is_in_iframe` as defined also in 431-gh-Automattic/tracks-events-registration at line 25 of `calypso_newsite_editor_launch_click.event.json`

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts (hunk 3 above)
Comment (yansern) [reply apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts]: Done. https://github.com/Automattic/wp-calypso/pull/49004/commits/30af0a4f1899c82679f2b3ca59c7123f5731fc43

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts (hunk 1 above)
Comment (yansern) [reply apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-button/index.ts]: Done. https://github.com/Automattic/wp-calypso/pull/49004/commits/471eed655eeabf91ed74b9746d5ee2f42fba3431

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/constants.ts [hunk 4]
```diff
@@ -1,10 +1,18 @@
...
-
+export const GUTENBOARDING_LAUNCH_FLOW = 'gutenboarding-launch';
+export const FOCUSED_LAUNCH_FLOW = 'focused-launch';
+export const SITE_LAUNCH_FLOW = 'launch-site';
 declare global {
 	interface Window {
 		wpcomEditorSiteLaunch?: {
+			launchUrl: string;
+			launchFlow: string;
+			isGutenboarding: boolean;
```
Comment (razvanpapadopol) [inline apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/constants.ts]: When logging `siteLaunchOptions` for a site created using `/start` I see that `isGutenboarding` is actually an empty string.
Screenshot 2021-01-27 at 09 04 07

Comment (yansern) [reply apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/constants.ts]: Looks like there's no way to use boolean (true/false) or number (1/0) via `wp_localize_script`, so we'll just have to agree that `""` is false and `"1"` is true.

In the consideration of whether to `"0"` or `""`, `""` would be preferable over `"0"` because `!""===true`  vs `!"0"===false`, so anyone doing `!isGutenboarding` would bring the right value.

I'll just update the type definition.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js (hunk 2 above)
Comment (yansern) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js:4]: Okay, scratch that. Changed my mind.
Now the `isGutenboarding` property either exists and hold the value "1" or does not exist at all.
The type definition is `isGutenboarding?: '1';`.
See https://github.com/Automattic/wp-calypso/pull/49004/commits/eeca2eb776c1b228f35958c3cc621dbe271278a4.

Comment (razvanpapadopol) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js:4]: I guess this last comment is related to https://github.com/Automattic/wp-calypso/pull/49004#discussion_r565238202 , right? However, the solution looks acceptable to me given the limitations 👌 .

**Comment by razvanpapadopol** [comment]
When loading `dev` version of ET to test this PR, more translations seem broken, other than the ones in production reported by @ciampo here: p1611740559105100/1611737961.102100-slack-C0KDTA48Y

I don't think this is something directly caused by this PR but since it was rebased to include the same code as ET has currently in production it's worth investigating or at least re-testing before releasing another ET version.

Examples:
Screenshot 2021-01-27 at 16 14 21
Screenshot 2021-01-27 at 16 14 27

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js (hunk 2 above)
Comment (yansern) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js:4]: LOL yes I replied in the wrong box.

File: apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/constants.ts (hunk 4 above)
Comment (yansern) [reply apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/constants.ts]: Reposting from https://github.com/Automattic/wp-calypso/pull/49004#discussion_r565265521 (as I initially commented in the wrong place).

Okay, scratch that. Changed my mind.
Now the `isGutenboarding` property either exists and hold the value "1" or does not exist at all.
The type definition is `isGutenboarding?: '1';`.
See https://github.com/Automattic/wp-calypso/pull/49004/commits/eeca2eb776c1b228f35958c3cc621dbe271278a4.

**Comment by yansern** [comment]
Looks like string issues are related to `import { __ } from '@wordpress/i18n'`. See p1611768666148500/1611737961.102100-slack-C0KDTA48Y

**Review by razvanpapadopol — APPROVED** [review]
All good from my side 👌
